pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -123,17 +154,24 @@ func NewLazyBinaryReader(
 		level.Debug(logger).Log("msg", "built index-header file", "path", path, "elapsed", time.Since(start))
 	}
 
-	return &LazyBinaryReader{
+	reader := &LazyBinaryReader{
 		logger:          logger,
 		filepath:        path,
 		metrics:         metrics,
-		usedAt:          atomic.NewInt64(time.Now().UnixNano()),
+		usedAt:          atomic.NewInt64(0),
 		onClosed:        onClosed,
 		readerFactory:   readerFactory,
 		blockID:         id,
 		lazyLoadingGate: lazyLoadingGate,
 		ctx:             ctx,
-	}, nil
+
+		loadedReaderC: make(chan readerRequest),
+		loadReqC:      make(chan loadRequest),
+		unloadReqC:    make(chan unloadRequest),
+	}
+	// TODO use a proper service lifecycler here to shut down this goroutine
```

Comment:
this came up in https://github.com/grafana/mimir/pull/8281#discussion_r1638174462 as well. Either @narqo or I will take this up to clean up the leaking goroutines soon (next week?)

Reply:
I feel that "a proper service lifecycler" would be an overkill here. Can we just pass a param in the unload request to return from the `controlLoop`, and pass it in the `Close()` call?

Reply:
Oh I see, how about having a `readerChannel atomic.Value` that holds the `chan readerRequest` and a `unloaded chan struct{}`

Then, when unloading we would do:
```
readerChannel.Set(nil) // or whatever the syntax is
close(done)
```

And `getReader` would look like:

```
func (r *LazyBinaryReader) getReader(ctx context.Context) loadedReader {
	readerReq := readerRequest{response: make(chan loadedReader)}
	select {
	case <-r.unloaded:
		return loadedReader{err: errUnloaded}
	case r.loadedReaderC <- readerReq:
		select {
		...
```

Reply:
how do you know to start the control loop again after it's been unloaded?

Reply:
Oh, sure, we'd have to this only when `Close()` is called, which is I assume when we're losing the reference here, is that right?

As I mentioned above:

> Can we just pass a param in the unload request to return from the controlLoop, and pass it in the Close() call?

Reply:
unfortunately no, from the docs of LazyBinaryReader: https://github.com/grafana/mimir/blob/268fb3cf7fe5bea002c7b0c344b0b5aa8780ac54/pkg/storegateway/indexheader/lazy_binary_reader.go#L177-L178